APEXMALHAR-1963 AbstractAsyncProcessor processes tuples asynchronously

chinmaykolhatkar opened the pull request:
Processes tuples async/parallel threads.
Also takes care of order and timeout of processing.
Also take care of checkpointing.

@gauravgopi123 Please review.

Commit:
APEXMALHAR-1963: AbstractAsyncProcessor processes tuples asynchronously

Added AbstractAsyncProcessor.
Fixing checkstyle violations.
Reducing logging for better performance.

sandeepdeshmukh commented on library/src/main/java/com/datatorrent/lib/async/AbstractAsyncProcessor.java:
```diff
@@ -0,0 +1,357 @@
+ *       processing by calling <i>enqueueTupleForProcessing</i>.</li>
+ *   <li>When tuple gets its turn of execution, <i>processTupleAsync</i> gets called from thread other than
+ *       operator thread.</li>
+ *   <li>Based on maintainTupleOrder parameter, the results are consolidated at end of every window and respective call
```
at _the_ end of

chinmaykolhatkar replied:
Done

sandeepdeshmukh commented on library/src/main/java/com/datatorrent/lib/async/AbstractAsyncProcessor.java:
```diff
@@ -0,0 +1,357 @@
+ * <ol>
+ *   <li>Tuple of type INPUT is received on input port.</li>
+ *   <li>Tuple gets enqueued into Executors by calling <i>enqueueTupleForProcessing</i>.
+ *       Once can override <i>processTuple</i> method in which case, its users responsibility to enqueue tuple for
```
Once -> One

chinmaykolhatkar replied:
Done

sandeepdeshmukh commented on library/src/main/java/com/datatorrent/lib/async/AbstractAsyncProcessor.java:
```diff
@@ -0,0 +1,357 @@
+ *   <li>Tuple gets enqueued into Executors by calling <i>enqueueTupleForProcessing</i>.
+ *       Once can override <i>processTuple</i> method in which case, its users responsibility to enqueue tuple for
+ *       processing by calling <i>enqueueTupleForProcessing</i>.</li>
+ *   <li>When tuple gets its turn of execution, <i>processTupleAsync</i> gets called from thread other than
```
-> When a tuple
Give some name to the other thread.

chinmaykolhatkar replied:
Done.

sandeepdeshmukh commented on library/src/main/java/com/datatorrent/lib/async/AbstractAsyncProcessor.java:
```diff
@@ -0,0 +1,357 @@
+  private static final Logger logger = LoggerFactory.getLogger(AbstractAsyncProcessor.class);
+
+  private transient ExecutorService executor;
+  private int numProcessors = 1;
```
Default of 1 for this purpose is not recommended. That will almost make it serial for each tuple. Suggest to increase it to a higher number. There is no perfect number for this. It will certainly depend on the system you are interacting with. 
Suggested number is 16 at least.

shubham-pathak22 replied:
Do we need to have max limit ?

chinmaykolhatkar replied:
Made it 16.

chinmaykolhatkar replied:
Actually let me keep this to 0. This means, the threads will be created as per need.

sandeepdeshmukh commented on library/src/main/java/com/datatorrent/lib/async/AbstractAsyncProcessor.java:
```diff
@@ -0,0 +1,357 @@
+      this.executor = Executors.newCachedThreadPool();
+    } else if (numProcessors > 0) {
+      this.executor = Executors.newFixedThreadPool(numProcessors);
+    } else {
```
@Min = 0  for numProcessors? Will save this else case.

chinmaykolhatkar replied:
Done.

shubham-pathak22 commented on library/src/main/java/com/datatorrent/lib/async/AbstractAsyncProcessor.java:
```diff
@@ -0,0 +1,357 @@
+   */
+  private Queue<ProcessTuple> waitingTuples = Lists.newLinkedList();
+
+  @InputPortFieldAnnotation(optional = true)
```
why optional ?

chinmaykolhatkar replied:
I'm actually considering to remove the input port form this.

I just need to ensure that this abstract operator is bound to operator lifecycle, so that the checkpointing can be handled properly.

The abstract has provided enqueue method which should take care of queing the tuples for processing.

shubham-pathak22 commented on library/src/main/java/com/datatorrent/lib/async/AbstractAsyncProcessor.java:
```diff
@@ -0,0 +1,357 @@
+    } else if (numProcessors == 0) {
+      this.executor = Executors.newCachedThreadPool();
+    } else if (numProcessors > 0) {
+      this.executor = Executors.newFixedThreadPool(numProcessors);
```
Do we need to distinguish between Executors.newSingleThreadExecutor() and Executors.newFixedThreadPool(numProcessors) with numProcessors = 1 ? Would it be functionally different ?

chinmaykolhatkar replied:
Done.

Commit:
Taking care of some review comments.
Removed input port as all this class needs to do is to bound to operator lifecycle for taking care of checkpointing and other work.
Updated unit tests for the same.

Updated comments.

Commit:
Taking care of review comments.
Updating the javadoc and some names of classes and variables.

gauravgopi123 commented on library/src/main/java/com/datatorrent/lib/async/AbstractAsyncProcessor.java:
```diff
@@ -0,0 +1,344 @@
+
+  /**
+   * Number of worker threads for processing
+   * 0   - Threads will be created as per need.
```
I don't think creating threads on need basis is good idea.. You need to have control on the number of threads

chinmaykolhatkar replied:
I agree.What would you suggest a good ma number will be. I'll remove the cached pool all together.
And put a limit on that.

chinmaykolhatkar replied:
So here is what I've done:
Min = 1 & Max = 32.
32 is the cap on number of threads enforced by javax Max annotation.

Default is set to 32.

Let me know if some other cap on number of threads need to be there.

gauravgopi123 commented on library/src/test/java/com/datatorrent/lib/async/AbstractAsyncProcessorTest.java:
```diff
@@ -0,0 +1,145 @@
+    for (int i = 0; i < 10; i++) {
+      Thread.sleep(waitInterval / 10);
+      async.endWindow();
+    }
```
what is this loop doing? endWindow() is called for every beginWindow and that has been done in previous loop (line 91-97)

chinmaykolhatkar replied:
Got it. I should have added the beginWindow also here. Will update.
This loop is just to make sure to wait for all the items to finish processing.

gauravgopi123 commented on library/src/test/java/com/datatorrent/lib/async/AbstractAsyncProcessorTest.java:
```diff
@@ -0,0 +1,145 @@
+      String outTuple1 = (String)outTuple;
+      String[] split = outTuple1.split(";");
+      Assert.assertEquals(2, split.length);
+      int current = Integer.parseInt(split[0]);
```
why not use swtich case?

chinmaykolhatkar replied:
Done.

Commit:
Taking care of review comments.
Adding limit on number of threads.

vrozov commented on library/src/main/java/com/datatorrent/lib/async/AbstractAsyncProcessor.java:
```diff
@@ -0,0 +1,341 @@
+    // Its left because its either not completed processing OR its not emitted because of ordering restrictions.
+    for (WorkItem item : waitingTuples) {
+      item.taskHandle = executor.submit(new Processor(item));
+      item.processState = State.SUBMITTED;
```
This possibly overwrites state of processed item.

chinmaykolhatkar replied:
calling executor.submit after variables are set. This should take care of overrides.

vrozov commented on library/src/main/java/com/datatorrent/lib/async/AbstractAsyncProcessor.java:
```diff
@@ -0,0 +1,341 @@
+    WorkItem item = new WorkItem(QUEUETUPLE);
+    waitingTuples.add(item);
+    item.taskHandle = executor.submit(new Processor(item));
+    item.processState = State.SUBMITTED;
```
This possibly overwrites state of processed item.

chinmaykolhatkar replied:
calling executor.submit after variables are set. This should take care of overrides.

Commit:
Taking care of possibility of overriding of states.
Making waitingTuples protected for subclass to access it.
